### PR TITLE
Integrate Polecat

### LINKED ISSUE #20246: Merge Polecat models into MOOSE
## Reason
The `LaRomance` that exist in MOOSE and Polecat have diverged. These need to be combined into the tensor_mechanics module into a single model.

## Design
Combine `LAROMANCEStressUpdateBase` in `tensor_mechanics` with the Polecat `LaRomance` model.

## Impact
Combined functionality for collaboration.

### PR DESCRIPTION
**Pull request by tophmatthews**
Closes #20246

- [x] Make work with unit converter
- [x] Make work with substepping
- [x] Make work with inplace tests
- [x] Make work with new verification tests
- [x] Make work with advanced models (not in this PR)

Thanks to @aruybalid for the initial code improvements to include partitioning.

### CODE REVIEW

**Comment by tophmatthews** [comment]
Something is missing with the substepping still so those tests are failing, but everything else seems good to go with the tests in place. I still need to test this with the other ROMS, so it's not done yet.

**Comment by cticenhour** [comment]
>The following file(s) are changed:
>
>  conda/libmesh/meta.yaml

@loganharbour is this notice going to show up every time there's a package change in next, even if the PR doesn't have an explicit change to this file in it? If so, is there a way we can limit the scope of this check to only the "actual" PR diff?

**Comment by tophmatthews** [comment]
Whew! I was trying to figure out what I did wrong there!

**Comment by loganharbour** [comment]
> > The following file(s) are changed:
> > conda/libmesh/meta.yaml
> 
> @loganharbour is this notice going to show up every time there's a package change in next, even if the PR doesn't have an explicit change to this file in it? If so, is there a way we can limit the scope of this check to only the "actual" PR diff?

This happens when someone rebases on next instead of devel and a conda change has not made it into devel. It's on my list to do git fu to probably get the merge base (which we use to determine the diffs), but it's not a priority because it doesn't happen often. I would suggest instead that people just rebase off of devel like we ask them.

**Comment by tophmatthews** [comment]
@recuero This new version of LaRomance has improved nonlinear convergence. Consequently, the tests you made to ensure that substepping was helping with the total error are not the same, and require a re-gold. The output shows that substepping with large timesteps approaches the value of no-substepping with small timesteps...I think that should be good enough for the tests? Can you verify I'm not missing something?

File: modules/tensor_mechanics/include/materials/LAROMANCEPartitionStressUpdateBase.h
Comment (tophmatthews) [inline modules/tensor_mechanics/include/materials/LAROMANCEPartitionStressUpdateBase.h:26]: I have to update this doco string still...

**Comment by recuero** [comment]
> the output shows that substepping with large timesteps approaches the value of no-substepping with small timesteps

Yeah, that's the main idea. I went through the updated csv files in this PR and the results make qualitatively sense to me. Usage of substepping still shows improved accuracy over regular stepping. The differences due to the updated stepping (`dt` changes) do not seem to affect the results a lot.

**Comment by tophmatthews** [comment]
@dschwen @bwspenc @lynnmunday Here is the Polecat integration we've been talking about for some time. It seems as though convergence is improved, although I'm not sure where or how...likely just a consequence of cleanups. Please run it with your current models (as we are) and ensure that similar results are achieved, and review the changes here. Thanks!

**Comment by tophmatthews** [comment]
But yes, it's just a swap. I sent you better verification tests and data though, so you'll want to use what I emailed you to build and test the proprietary model.

**Comment by tophmatthews** [comment]
> @tophathews,

Coffee still waiting to kick in?

**Comment by dschwen** [comment]
> > @tophathews,
> 
> Coffee still waiting to kick in?

🎩 ⛏️  That's not your name?

**Comment by tophmatthews** [comment]
Oh, maybe zip problems...I'll a different way.

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (tophmatthews) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:261]: FYI @dschwen this is required (return true led to errors?)

Comment (dschwen) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:261]: Got it. That also means the function name is good. Not how I would have designed it though, as the parameter is added by a parent class and I prefer parameter handling to happen at the same level.

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (tophmatthews) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:1191]: This fixed up the `nan`'s I was seeing @dschwen , not sure if it was the same for you or similar to your fix?

**Comment by tophmatthews** [comment]
Hey @dschwen and @bwspenc , there are a couple more things I need to do to stamp this integration as done, but I think it'll be easier to get this in. @dschwen already showed that answers are the same as previous implementations, but convergence has a bit of problem in certain inputs. There are some fixes there that we are working on, but I don't want to get this tangled any more. Can we merge this?

**Comment by dschwen** [comment]
Yeah, that's fine with me

**Comment by tophmatthews** [comment]
Actually...hold off on the merge, I found the bug @dschwen ... I think 🤞

**Comment by tophmatthews** [comment]
Alright, give it a try @dschwen

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (tophmatthews) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:781]: 🙄 
(In case you were curious @dschwen )

**Comment by dschwen** [comment]
Running this right now and it looks pretty good! Let me do some more checks with cranked up loading.

File: modules/tensor_mechanics/src/materials/LAROMANCEPartitionStressUpdateBase.C
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/LAROMANCEPartitionStressUpdateBase.C:39]: We tend to use

```suggestion
  for (const auto i: index_range(_input_values))
```

in new code these days. (`#include "libmesh/int_range.h"`)

**Review by dschwen — COMMENTED** [review]
Works perfectly now!

File: modules/tensor_mechanics/src/materials/LAROMANCEPartitionStressUpdateBase.C
Comment (dschwen) [reply modules/tensor_mechanics/src/materials/LAROMANCEPartitionStressUpdateBase.C:39]: Sure we do that plenty. Here you seem to need the index though.

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/LAROMANCEPartitionStressUpdateBase.C:39]: Yeah, totally. Just curious. Is this a fix it now or fix it later suggestion here?

Comment (dschwen) [reply modules/tensor_mechanics/src/materials/LAROMANCEPartitionStressUpdateBase.C:39]: Up to you. I can approve this now.

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (lynnmunday) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:639]: should this set weights=0?  Same with line 670.  Is USELIMIT the same as DONOTHING enum you have commented out?

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:639]: If USELIMIT is used, it just changes the input limit to the edge of the box of applicability, and runs it normally. DONOTHING sets weight to zero.

Comment (lynnmunday) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:639]: what does USELIMIT set the weights to?  I don't see where weights would be set.

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:639]: Essentially multiplies by 1. You can't set the weights to anything, except zero, only multiple them by the local weight, since it's the accumulation of weights for all input values.

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:333]:
```suggestion
    _global_limits[i]= {std::numeric_limits<Real>::max(), 0.0};
```
?

**Comment by dschwen** [comment]
FYI

|phase|NEML    | old LaRomance | new LaRomance |
|--------|-----------|-----------------------|------------------------|
|1|2072s|5042s|8281s|
|2|1331s|7787s|24612s|

**Comment by tophmatthews** [comment]
There is one more fix that @aruybalid has queued up that may help that @dschwen . Can we get this in so he can apply some new patches? After a good review from you and @lynnmunday of course.

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:333]: Yes, that's nicer...next round?

**Review by dschwen — DISMISSED** [review]

**Comment by lynnmunday** [comment]
@tophmatthews I approve.  I don't think I have an approval button I can press though.

**Review by lynnmunday — DISMISSED** [review]

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (lynnmunday) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:86]: you can fix that when you make your next patch.

**Comment by tophmatthews** [comment]
Oh wow...Profiling pointed out some dummy mistakes...I sped it up 90% on a dummy 1 element problem...standby

**Review by bwspenc — COMMENTED** [review]
I looked through this, and the code looks clean. Just one question, though: based on the naming of the base classes, I would have expected everything related to the partitioned model to be in LAROMANCEPartitionStressUpdateBase, but there's an awful lot of partition-related code in LAROMANCEStressUpdateBase. Just curious what the thinking is there.

**Comment by tophmatthews** [comment]
> but there's an awful lot of partition-related code in LAROMANCEStressUpdateBase

It's actually not a bad point that the `computePartitionWeights` should just return `1` in `LAROMANCEStressUpdateBase`, and be overriden in `LAROMANCEPartitionStressUpdateBase`. That way `computeSecondPartitionWeight` and `computeDSecondPartitionWeightDStress` could only be defined in `LAROMANCEPartitionStressUpdateBase` as well.

As for the other parts, most of it is the required extra dimension on the vectors, so I think that needs to stay. I clean up those pieces though, thanks for the tip @bwspenc

**Comment by aruybalid** [comment]
Hi guys, sorry to be a bit of noob commenting in Github, so I am not sure if this is the correct way and place to provide comments/feedback. I was running some tests and ran into the error:
```
*** ERROR ***
In rom_stress_prediction_base: input parameter with value (5.2476e+12) is out of upper global range (2.78853e+13)
```
which is weird, because that value is not the upper limit. I think a copy/paste error slipped in the `computeTileWeight` method in `LAROMANCEStressUpdateBase.C`:

```
      // If input is above the upper tile limit
      else if (input > _input_limits[p][t][in_index][1])
      {
        if (_input_limits[p][t][in_index][1] == _global_limits[in_index].second)
        {
          if (_window_failure[in_index].second == WindowFailure::EXTRAPOLATE)
            mooseError("Internal error. Extrapolate not appropriate for upper bound");
          else if (_window_failure[in_index].second == WindowFailure::USELIMIT)
            input = _input_limits[p][t][in_index][1];
          else
          {
            // Set input to edge of limit so it is found later in computePartitionWeights
            input = _input_limits[p][t][in_index][0];
            weights[p][t] = 0.0;
            std::stringstream msg;
            msg << "In " << _name << ": input parameter with value ("
                << MetaPhysicL::raw_value(input) << ") is out of upper global range ("
                << _input_limits[p][t][in_index][1] << ")";

            if (_window_failure[in_index].first == WindowFailure::WARN)
              mooseWarning(msg.str());
            else if (_window_failure[in_index].first == WindowFailure::ERROR)
              mooseError(msg.str());
            else if (_window_failure[in_index].first == WindowFailure::EXCEPTION)
              mooseException(msg.str());

```
I believe it should be:
```
      // If input is above the upper tile limit
      else if (input > _input_limits[p][t][in_index][1])
      {
        if (_input_limits[p][t][in_index][1] == _global_limits[in_index].second)
        {
          if (_window_failure[in_index].second == WindowFailure::EXTRAPOLATE)
            mooseError("Internal error. Extrapolate not appropriate for upper bound");
          else if (_window_failure[in_index].second == WindowFailure::USELIMIT)
            input = _input_limits[p][t][in_index][1];
          else
          {
            // Set input to edge of limit so it is found later in computePartitionWeights
            input = _input_limits[p][t][in_index][1];
            weights[p][t] = 0.0;
            std::stringstream msg;
            msg << "In " << _name << ": input parameter with value ("
                << MetaPhysicL::raw_value(input) << ") is out of upper global range ("
                << _input_limits[p][t][in_index][1] << ")";

            if (_window_failure[in_index].second == WindowFailure::WARN)
              mooseWarning(msg.str());
            else if (_window_failure[in_index].second == WindowFailure::ERROR)
              mooseError(msg.str());
            else if (_window_failure[in_index].second == WindowFailure::EXCEPTION)
              mooseException(msg.str());
```

**Comment by tophmatthews** [comment]
Nice catch! Next time just comment directly in the spot in the code so I don't have to look through 30 lines to spot-the-difference 😄

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (aruybalid) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:671]: `input = _input_limits[p][t][in_index][1];`

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (aruybalid) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:678]: `if (_window_failure[in_index].second == WindowFailure::WARN)`

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:671]: 🤘

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (aruybalid) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:680]: `else if (_window_failure[in_index].second == WindowFailure::ERROR)`

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (aruybalid) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:682]: `else if (_window_failure[in_index].first == WindowFailure::EXCEPTION)`

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:680]: Oh I see there was a lot...

File: modules/tensor_mechanics/include/materials/LAROMANCEPartitionStressUpdateBase.h
Comment (tophmatthews) [inline modules/tensor_mechanics/include/materials/LAROMANCEPartitionStressUpdateBase.h:49]: This will require redefinition of this method in the derivative classes.

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (aruybalid) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:675]: print the input before it was reset to limit:
```
msg << "In " << _name << ": input parameter with value ("
                << input_before_set_to_limit << ") is out of upper global range ("
                << _input_limits[p][t][in_index][1] << ")";
```

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:675]: Good catch!

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (aruybalid) [inline modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:671]: store value of input before setting it to limit value (for later console print message):
```
GenericReal<is_ad> input_before_set_to_limit = input;
input = _input_limits[p][t][in_index][1];
```

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:671]: No, I don't think I need that, I can just do `input = _input_limits[p][t][in_index][1]` after I make the msg, no?

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (aruybalid) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:675]: Do the same for the lower limit: lines 643 and 639 (for some reason I don't get the little + symbol to add a comment in those lines?)

File: modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C
Comment (aruybalid) [reply modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C]: jep, that works. Also in the lower limit case..

**Review by dschwen — APPROVED** [review]

**Comment by tophmatthews** [comment]
We're in (what feels like) a narrow window where moose/next is green so this test is passing, is this good to go in @dschwen @bwspenc ?

**Review by bwspenc — APPROVED** [review]

**Pull request merged**
